### PR TITLE
feat(off): add off feature

### PR DESCRIPTION
**Pull request by hekike**
Add `.off(method, path)` method.

## Related issues

- https://github.com/delvedor/find-my-way/issues/40

### CODE REVIEW

**Comment by delvedor** [comment]
Hi! Thank you for the proposal!

I think the best option would be regenerate the entire tree every time you call `.off`.
1. No need to change the internals which are highly optimized.
2. add the parent node to a Node means that you are creating a circular reference, in this way you can't print the tree (`JSON.stringify(router.tree, null, 2)`) which is very useful for debugging.

We can create an array where we save all the routes and if the user calls `.off` we remove the specific route from that array and then we regenerate the tree.
What do you think?

**Comment by hekike** [comment]
@delvedor thanks for the feedback. I'm okay with the solution what you proposed.
Do you want to work on it or should I go ahead?

**Comment by delvedor** [comment]
@hekike go ahead! :)

**Comment by hekike** [comment]
@delvedor done, how about now? ;)

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "find-my-way",
-  "version": "1.7.0",
+  "version": "1.7.1",
```
Comment (mcollina) [inline package.json]: Why this?

File: index.js [hunk 2]
```diff
@@ -183,6 +190,38 @@ Router.prototype._insert = function _insert (method, path, kind, params, handler
...
+    })
+  }
+
+  // method validation
+  assert(typeof method === 'string', 'Method should be a string')
+  assert(httpMethods.indexOf(method) !== -1, `Method '${method}' is not an http method.`)
+  // path validation
+  assert(typeof path === 'string', 'Path should be a string')
+  assert(path.length > 0, 'The path could not be empty')
+  assert(path[0] === '/' || path[0] === '*', 'The first character of a path should be `/` or `*`')
```
Comment (mcollina) [inline index.js]: it seems a part of this check is not really tested.

File: README.md [hunk 3]
```diff
@@ -117,6 +117,31 @@ parametric(regex)
...
+// => [{ handler: Function, params: Object, store: Object}]
+// => null
+```
+
+<a name="reset"></a>
+#### reset()
+Empty router.
+```js
+router.reset()
+```
```
Comment (mcollina) [inline README.md]: can you add a test for reset()?

**Comment by delvedor** [comment]
Try add this test:
```js
test('off with nested wildcards with parametric and static', t => {
  t.plan(2)
  const findMyWay = FindMyWay({
    defaultRoute: (req, res) => {
      t.fail('we should not be here, the url is: ' + req.url)
    }
  })

  findMyWay.on('GET', '*', (req, res, params) => {
    t.is(params['*'], '/foo2/first/second')
  })
  findMyWay.on('GET', '/foo1/*', () => {})
  findMyWay.on('GET', '/foo2/*', () => {})
  findMyWay.on('GET', '/foo3/:param', () => {})
  findMyWay.on('GET', '/foo3/*', () => {})
  findMyWay.on('GET', '/foo4/param/hello/test/long/route', () => {})

  findMyWay.off('GET', '/foo3/*')
  t.notOk(findMyWay.find('GET', '/foo3/first/second'))

  findMyWay.off('GET', '/foo2/*')
  findMyWay.lookup(
    { method: 'GET', url: '/foo2/first/second' },
    null
  )
})
```
For some reason is not passing on my machine.

**Comment by hekike** [comment]
@delvedor are you sure it's related to the `.off`, for me, it's failing any `off` call.

```js
test('off with nested wildcards with parametric and static', t => {
  t.plan(1)
  const findMyWay = FindMyWay({
    defaultRoute: (req, res) => {
      t.fail('we should not be here, the url is: ' + req.url)
    }
  })

  findMyWay.on('GET', '*', (req, res, params) => {
    t.is(params['*'], '/foo2/first/second')
  })
  findMyWay.on('GET', '/foo1/*', () => {})
  findMyWay.on('GET', '/foo2/*', () => {})
  findMyWay.on('GET', '/foo3/:param', () => {})
  findMyWay.on('GET', '/foo3/*', () => {})
  findMyWay.on('GET', '/foo4/param/hello/test/long/route', () => {})

  // findMyWay.off('GET', '/foo3/*')
  // t.notOk(findMyWay.find('GET', '/foo3/first/second'))

  // findMyWay.off('GET', '/foo2/*')
  findMyWay.lookup(
    { method: 'GET', url: '/foo2/first/second' },
    null
  )
})
```

**Comment by delvedor** [comment]
@hekike this is the standard situation:
```js
test('check 1', t => {
  t.plan(1)
  const findMyWay = FindMyWay({
    defaultRoute: (req, res) => {
      t.fail('we should not be here, the url is: ' + req.url)
    }
  })

  findMyWay.on('GET', '*', () => {})
  findMyWay.on('GET', '/foo1/*', () => {})
  findMyWay.on('GET', '/foo2/*', (req, res, params) => {
    t.is(params['*'], 'first/second')
  })
  findMyWay.on('GET', '/foo3/:param', () => {})
  findMyWay.on('GET', '/foo3/*', () => {})
  findMyWay.on('GET', '/foo4/param/hello/test/long/route', () => {})

  findMyWay.lookup(
    { method: 'GET', url: '/foo2/first/second' },
    null
  )
})
```

And this is the one without `/foo2/*`:
```js
test('check 2', t => {
  t.plan(1)
  const findMyWay = FindMyWay({
    defaultRoute: (req, res) => {
      t.fail('we should not be here, the url is: ' + req.url)
    }
  })

  findMyWay.on('GET', '*', (req, res, params) => {
    t.is(params['*'], '/foo2/first/second')
  })
  findMyWay.on('GET', '/foo1/*', () => {})
  findMyWay.on('GET', '/foo3/:param', () => {})
  findMyWay.on('GET', '/foo3/*', () => {})
  findMyWay.on('GET', '/foo4/param/hello/test/long/route', () => {})

  findMyWay.lookup(
    { method: 'GET', url: '/foo2/first/second' },
    null
  )
})
```
Both tests are passing on my machine, this means that `off` is messing up something :S

**Comment by hekike** [comment]
@delvedor I found the issue and fixed it. Thanks for the help!

**Comment by hekike** [comment]
@delvedor I also added your wildcard test case to the tests.

**Comment by delvedor** [comment]
@hekike you are welcome!
For the sake of curiosity, what did you changed?

File: test/methods.test.js [hunk 4]
```diff
@@ -37,6 +67,124 @@ test('register a route with multiple methods', t => {
...
+    t.is(params['*'], '/foo2/first/second')
+  })
+  findMyWay.on('GET', '/foo1/*', () => {})
+  findMyWay.on('GET', '/foo2/*', () => {})
+  findMyWay.on('GET', '/foo3/:param', () => {})
+  findMyWay.on('GET', '/foo3/*', () => {})
+  findMyWay.on('GET', '/foo4/param/hello/test/long/route', () => {})
+
+  var route1 = findMyWay.find('GET', '/foo2/first/second')
+  t.ok(route1.params['*'])
```
Comment (delvedor) [inline test/methods.test.js]: Can you check also the value of the wildcard?
eg:
```js
t.is(route1.params['*'], 'first/second')
```

Same below :)

**Comment by hekike** [comment]
@delvedor I added the param checks to the tests.
The issue was earlier that I pushed the route to late some return run before.

**Comment by delvedor** [comment]
Landed in [`v1.8.0`](https://github.com/delvedor/find-my-way/releases/tag/v1.8.0) :)

**Comment by hekike** [comment]
@delvedor thanks! 🎉

**Pull request merged**